### PR TITLE
Use ScenarioRunner.submit in Daml Script

### PR DESCRIPTION
**Pull request by cocreature**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/tests/src/DA/Test/ScriptService.hs [hunk 1]
```diff
@@ -101,7 +101,7 @@ main =
                       "      do pure v",
                       "testCreate = do",
                       "  p <- allocateParty \"p\"",
-                      "  submit p $ createCmd (T p 42)",
+                      "  submit p $ createCmd (T p 42)" ,
```
Comment (remyhaemmerle-da) [inline compiler/damlc/tests/src/DA/Test/ScriptService.hs:104]: ?

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala [hunk 2]
```diff
@@ -278,15 +224,14 @@ class IdeLedgerClient(val compiledPackages: CompiledPackages) extends ScriptLedg
   )(implicit ec: ExecutionContext, mat: Materializer): Future[Either[Unit, Unit]] = {
     unsafeSubmit(actAs, readAs, commands, optLocation)
       .map({
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala:226]:
```suggestion
      .map {
```

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala [hunk 3]
```diff
@@ -278,15 +224,14 @@ class IdeLedgerClient(val compiledPackages: CompiledPackages) extends ScriptLedg
...
-        Future.successful(Right(()))
+        case _: ScenarioRunner.Commit[_] => Left(())
+        case error: ScenarioRunner.SubmissionError =>
+          ledger = ledger.insertAssertMustFail(actAs.toSet, readAs, optLocation)
+          seed = ScenarioRunner.nextSeed(
+            error.ptx.unwind.context.nextActionChildSeed
+          )
+          clearPtx()
+          Right(())
       })
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala:235]:
```suggestion
       }
```

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala [hunk 4]
```diff
@@ -9,84 +9,88 @@ import com.daml.ledger.api.domain.PartyDetails
...
   )
-  val onLedger = machine.ledgerMode match {
-    case OffLedger => throw SRequiresOnLedger("ScenarioRunner")
-    case onLedger: OnLedger => onLedger
-  }
-  val scenarioRunner = ScenarioRunner(machine, seed)
+
+  private[this] val preprocessor = new engine.preprocessing.CommandPreprocessor(compiledPackages)
+
+  var ledger: ScenarioLedger = ScenarioLedger.initialLedger(Time.Timestamp.Epoch)
```
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala:83]: Do we ever modified this variable outside the class ? 
If not I will prefer to not expose the variable part of this this field 
```suggestion
  def ledger: ScenarioLedger = ledger_
  private[this] var ledger_: ScenarioLedger = ScenarioLedger.initialLedger(Time.Timestamp.Epoch)
```

**Review by remyhaemmerle-da — APPROVED** [review]
Look much nicer.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala (hunk 4 above)
Comment (cocreature) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala]: cd51ee542379fa3efc11a65cca6f387ebcb7b8db

File: compiler/damlc/tests/src/DA/Test/ScriptService.hs (hunk 1 above)
Comment (cocreature) [reply compiler/damlc/tests/src/DA/Test/ScriptService.hs]: 8c67d2a28f5f97c105f731301f7072bd1c4cc0d8 not sure how that snuck in

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala [hunk 5]
```diff
@@ -9,84 +9,89 @@ import com.daml.ledger.api.domain.PartyDetails
...
-  val onLedger = machine.ledgerMode match {
-    case OffLedger => throw SRequiresOnLedger("ScenarioRunner")
-    case onLedger: OnLedger => onLedger
-  }
-  val scenarioRunner = ScenarioRunner(machine, seed)
+
+  private[this] val preprocessor = new engine.preprocessing.CommandPreprocessor(compiledPackages)
+
+  private var _ledger: ScenarioLedger = ScenarioLedger.initialLedger(Time.Timestamp.Epoch)
+  def ledger: ScenarioLedger = _ledger
```
Comment (stefanobaghino-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala:84]: :trollface: 
```suggestion
  private var ledger: ScenarioLedger = ScenarioLedger.initialLedger(Time.Timestamp.Epoch)
  def getLedger: ScenarioLedger = ledger
```
Just trolling, please don't do this.

Comment (cocreature) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/IdeLedgerClient.scala:84]: working on adding a lombok dependency to autogneerate this

**Pull request merged**